authorize_helper.go

```diff
@@ -41,11 +41,15 @@ import (
 func GetRedirectURIFromRequestValues(values url.Values) (string, error) {
 	// rfc6749 3.1.   Authorization Endpoint
 	// The endpoint URI MAY include an "application/x-www-form-urlencoded" formatted (per Appendix B) query component
-	redirectURI, err := url.QueryUnescape(values.Get("redirect_uri"))
+	rawRedirectURI := values.Get("redirect_uri")
+	redirectURI, err := url.Parse(rawRedirectURI)
 	if err != nil {
 		return "", errors.WithStack(ErrInvalidRequest.WithHint(`The "redirect_uri" parameter is malformed or missing.`).WithCause(err).WithDebug(err.Error()))
+	} else if rawRedirectURI != "" && (redirectURI.Scheme == "" || redirectURI.Host == "") {
```

Comment:
Any particular reason to add this additional check at this point?

Reply:
1. According to the [test case](https://github.com/ory/fosite/blob/b53f8f58f0b9889d044cf9a8e2604316f0559ff6/authorize_helper_test.go#L62) empty `redirect_uri` is allowed

2. Golang url.parse doesn't catch [bad URL scheme or host](https://stackoverflow.com/a/55551215) thus the method fails to catch urls like `https//google.com/foo=bar`

Reply:
That's why we have `url.ParseRequestURI("https//google.com/foo=bar")` :)

https://play.golang.org/p/eblL4ikZTSi

Reply:
Please keep in mind that this errors if the string is empty, so you need to handle that as a special case!

Reply:
But this function should not be doing any validation based on its docstring!

So see other comments. Both @ajanthan and me think now that this function should just be removed and replaced with simply calling `values.Get("redirect_uri")`. The validation happens anyway in `MatchRedirectURIWithClientRedirectURIs`.

Reply:
Right, overlooked that - makes sense!